test/integration/filters/modify_buffer_filter.cc

```diff
@@ -0,0 +1,57 @@
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/filter_config.h"
+
+#include "extensions/filters/http/common/empty_http_filter_config.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+
+// A filter that buffers the entire request/response, then doubles
+// the content of the filter buffer.
+class ModifyBufferStreamFilter : public Http::PassThroughFilter {
+public:
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool end_stream) {
+    if (end_stream) {
+      decoder_callbacks_->modifyDecodingBuffer([](auto& buffer) {
```

Comment:
As you call out in the integration test, this doesn't modify the entire request body when the data arrives with end stream.  I suspect this generally won't be what the user wants - we've seen plenty of these bugs (e.g.  #5674) where we need to
callbacks_->addDecodedData(data, true);
to capture data sent inline with end_stream

This means that the reference implementation of this function doesn't do what it says (and folks may copy it without realizing) and also that the design is subject to the data-with-final-end-stream pattern which I'd really love to avoid.  Can we try to fix this?